torch_geometric/nn/models/revgnn.py

```diff
@@ -0,0 +1,287 @@
+        filtered_detached_inputs = tuple(
+            filter(
+                lambda x: x.requires_grad
+                if isinstance(x, torch.Tensor) else False,
+                detached_inputs,
+            ))
+        gradients = torch.autograd.grad(
+            outputs=temp_output,
+            inputs=filtered_detached_inputs + ctx.weights,
+            grad_outputs=grad_outputs,
+        )
+
+        input_gradients = []
+        i = 0
+        for rg in ctx.input_requires_grad:
+            if rg:
+                input_gradients.append(gradients[i])
+                i += 1
+            else:
+                input_gradients.append(None)
+
+        gradients = tuple(input_gradients) + gradients[-len(ctx.weights):]
+
+        return (None, None, None, None) + gradients
+
+
+class InvertibleModule(nn.Module, ABC):
+    r"""An abstract class for implementing Invertible Modules.
+
+    Args:
+        disable (bool, optional): This disables using the InvertibleFunction.
+            Therefore, it executes functions without memory savings.
+        num_bwd_passes (int, optional):
+            Number of backward passes to retain a link with the output.
+            After the last backward pass the output is discarded
+            and memory is freed. (default: :obj:`1`)
+    """
+    def __init__(self, disable=False, num_bwd_passes=1):
+        super().__init__()
+        self.disable = disable
+        self.num_bwd_passes = num_bwd_passes
+
+    def forward(self, *args, **kwargs):
+        args = (x, edge_index) + args
+        y = self._fn_apply(args, self._forward, self._inverse)
+        return y
+
+    def inverse(self, *args, **kwargs):
+        args = (y, edge_index) + args
+        x = self._fn_apply(args, self._inverse, self._forward)
+        return x
+
+    @abstractmethod
+    def _forward(self):
+        pass
+
+    @abstractmethod
+    def _inverse(self):
+        pass
+
+    def _fn_apply(self, args, fn, fn_inverse):
+        if not self.disable:
+            y = InvertibleFunction.apply(
+                fn,
+                fn_inverse,
+                self.num_bwd_passes,
+                len(args),
+                *(args + tuple(p
+                               for p in self.parameters() if p.requires_grad)),
+            )
+        else:
+            y = fn(*args)
+
+        return self.if_unpack(y)
+
+    @staticmethod
+    def if_unpack(out):
+        # If the layer only has one input, we unpack the tuple again
+        if isinstance(out, tuple) and len(out) == 1:
+            return out[0]
+        return out
+
+
+class GroupAddRev(InvertibleModule):
+    r"""InvertibleModule of Grouped Reversible GNNs for saving GPU memory.
+    Proposed in Training Graph Neural Networks with 1000 Layers:
+    https://arxiv.org/abs/2106.07476.
+    This module enables training arbitary deep GNNs with memory complexity
+    that is independent of the number of layers.
+
+    Args:
+        gnn (Union[nn.Module, nn.ModuleList]):
```

Comment:
Here do you expect a GNN layer or a full GNN network? Perhaps `gnn_layer` or just `layer` is a less confusing name

Reply:
We can expect any gnn block with forward function input as `x` and `edge_index` .